skyportal/handlers/api/internal/profile.py

```diff
@@ -89,20 +113,68 @@ def put(self):
                 schema: Error
         """
         data = self.get_json()
+
+        user = (User.query.filter(User.username == self.current_user.username)
+                    .first())
+
         if 'preferences' not in data:
-            return self.error('Invalid request body: missing required "preferences" parameter')
+            return self.error(
+                              'Invalid request body: missing required '
+                              '"preferences" parameter'
+                             )
         preferences = data['preferences']
+        if preferences.get("first_name") is not None:
+            user.first_name = preferences.pop("first_name")
+
+        if preferences.get("last_name") is not None:
+            user.last_name = preferences.pop("last_name")
+
+        if preferences.get("contact_phone") is not None:
+            phone = preferences.pop("contact_phone")
+            if phone not in [None, ""]:
+                try:
+                    if not phonenumbers.is_possible_number(
+                            phonenumbers.parse(phone, "US")):
+                        return self.error(
+                                  'Phone number given is not valid'
+                                 )
+                except NumberParseException:
+                    return self.error(
+                                  'Could not parse input as a phone number'
+                                 )
+                user.contact_phone = phone
+            else:
+                user.contact_phone = None
+
+        if preferences.get("contact_email") is not None:
+            email = preferences.pop("contact_email")
+            if email not in [None, ""]:
+                if not validate_email(
+                       email_address=email,
+                        check_regex=True,
+                        check_mx=False,
+                        use_blacklist=True,
+                        debug=False):
+                    return self.error(
+                              'Email does not appear to be valid'
+                             )
+                user.contact_email = email
+            else:
+                user.contact_email = None
+
         # Do not save blank fields (empty strings)
         for k, v in preferences.items():
             if isinstance(v, dict):
-                preferences[k] = {key: val for key, val in v.items() if val != ''}
-        user_prefs = deepcopy(self.current_user.preferences)
+                preferences[k] = \
```

Comment:
88 is the new 79/80, so this can be reverted (and using the curly braces in this expression to span two lines, or introducing parentheses in other cases, is preferred over trailing slashes)